pyodide-build/pyodide_build/out_of_tree/pypi.py

```diff
@@ -0,0 +1,386 @@
+import json
+import os
+import shutil
+import sys
+import tempfile
+from collections.abc import Generator
+from contextlib import contextmanager
+from email.message import EmailMessage
+from email.parser import BytesParser
+from io import BytesIO
+from operator import attrgetter
+from pathlib import Path
+from platform import python_version
+from typing import TYPE_CHECKING, Any, BinaryIO, cast
+from urllib.parse import urlparse
+from zipfile import ZipFile
+
+import requests
+from packaging.requirements import Requirement
+from packaging.utils import canonicalize_name
+from packaging.version import Version
+from resolvelib import BaseReporter, Resolver
+from resolvelib.providers import AbstractProvider
+from unearth.evaluator import TargetPython
+from unearth.finder import PackageFinder
+
+from .. import common
+from . import build
+
+
+@contextmanager
+def stream_redirected(to=os.devnull, stream=None):
+    if stream is None:
+        stream = sys.stdout
+    stream_fd = stream.fileno()
+    if type(to) == str:
+        to = open(to, "w")
+    with os.fdopen(os.dup(stream_fd), "wb") as copied:
+        stream.flush()
+        os.dup2(to.fileno(), stream_fd)  # $ exec >&to
+        try:
+            yield stream  # allow code to be run with the redirected stream
+        finally:
+            # restore stream to its previous value
+            # NOTE: dup2 makes stream_fd inheritable unconditionally
+            stream.flush()
+            os.dup2(copied.fileno(), stream_fd)  # $ exec >&copied
+            to = None
+
+
+# cache of built sdists by url source to wheel file path
+BUILD_CACHE: dict[str, dict[str, Any]] = {}
+
+
+def get_cached_built_wheel(url):
+    parsed_url = urlparse(url)
+    gz_name = Path(parsed_url.path).name
+    if url in BUILD_CACHE:
+        return BUILD_CACHE[url]["path"]
+    else:
+        cache_entry = {}
+        build_dir = tempfile.TemporaryDirectory()
+        cache_entry["build_dir"] = build_dir
+        os.chdir(build_dir.name)
+        with tempfile.NamedTemporaryFile(suffix=".tar.gz", delete=False) as f:
+            data = requests.get(url).content
+            f.write(data)
+            f.close()
+            shutil.unpack_archive(f.name, build_dir.name)
+            os.unlink(f.name)
+            files = list(os.listdir(build_dir.name))
+            if len(files) == 1:
+                os.chdir(os.path.join(build_dir.name, files[0]))
+            else:
+                os.chdir(build_dir.name)
```

Comment:
Why not:
```suggestion        
        data = requests.get(url).content
        with tempfile.NamedTemporaryFile(suffix=".tar.gz") as f:
            f.write(data)
            f.close()
            shutil.unpack_archive(f.name, build_dir.name)
        files = list(os.listdir(build_dir.name))
        if len(files) == 1:
            os.chdir(os.path.join(build_dir.name, files[0]))
        else:
            os.chdir(build_dir.name)
```

Reply:
because NamedTemporaryFile deletes on first close, not when the context manager exits. Stupid, but it is how it works annoyingly.